Can't load_from_checkpoint with QuantizationAwareTraining callback during training

for PL v1.2.0

```py
LightningModel.load_from_checkpoint(str(ckpt_path))
```
gives
```
  File "//checkpoint/utils.py", line 11, in load_pl_model
    pl_model: LightningModel = LightningModel.load_from_checkpoint(str(ckpt_path))
  File "/.venv/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 156, in load_from_checkpoint
    model = cls._load_model_state(checkpoint, strict=strict, **kwargs)
  File "//.venv/lib/python3.8/site-packages/pytorch_lightning/core/saving.py", line 204, in _load_model_state
    model.load_state_dict(checkpoint['state_dict'], strict=strict)
  File "//.venv/lib/python3.8/site-packages/torch/nn/modules/module.py", line 1223, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for LightningModel:
Unexpected key(s) in state_dict: "quant.activation_post_process.fake_quant_enabled", "quant.activation_post_process.observer_enabled", "quant.activation_post_process.scale", "quant.activation_post_process.zero_point", "quant.activation_post_process.activation_post_process.eps", "quant.activation_post_process.activation_post_process.min_val", "quant.activation_post_process.activation_post_process.max_val", "model.encoder._bn0.activation_post_process.fake_quant_enabled", "model.encoder._bn0.activation_post_process.observer_enabled", "model.encoder._bn0.activation_post_process.scale", "model.encoder._bn0.activation_post_process.zero_point", .............
```

> > well you can load it without strict, see
> 
> In that case, we're just loading an un-quantized model, right? What if users do want to load the in8 weights?
> 
> For example, there are post training steps (e.g. adding a wrapper to output tuple instead of dict) that happen on another machine. Trained checkpoint has to be able to load elsewhere.

well, no checkpoint contains quantized weights, as you can see in the callback, the quantization is done to the trained model when training ends compare to checkpoints which are saved from during training before the quantization conversion is called...
so if you want to work with the quantized model, you need to save it extra after training but that seems to be not working as the wrapped forward method is not pickable...
```
>       pickler.dump(obj)
E       _pickle.PicklingError: Can't pickle <function RegressionModel.forward at 0x7f7202cb48b0>: it's not the same object as tests.helpers.simple_models.RegressionModel.forward
```
so you need to use  `QuantizationAwareTraining(input_compatible=False)`

> > this design requires preparing a model which inherits from the QuantizationMixin so as a user you can't simply append/deploy quantization for any model you want... right?
> 
> Yup, I don't want to impose this extra constraint, but I couldn't figure out a better way...

I see, but as I just mentioned the checkpoint does not include any quantized weights, just collected statistics to run the conversion "again"

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -177,7 +177,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Enabled traditional/manual launching of DDP processes through `LOCAL_RANK` and `NODE_RANK` environment variable assignments ([#7480](https://github.com/PyTorchLightning/pytorch-lightning/pull/7480))
 
 
-- Added `convert_on_fit_end` argument to `QuantizationAwareTraining` ([#8464](https://github.com/PyTorchLightning/pytorch-lightning/pull/8464))
+- Added `quantize_on_fit_end` argument to `QuantizationAwareTraining` ([#8464](https://github.com/PyTorchLightning/pytorch-lightning/pull/8464))
 
 
 - Added experimental support for loop specialization ([#8226](https://github.com/PyTorchLightning/pytorch-lightning/pull/8226))

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -175,6 +175,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Enabled traditional/manual launching of DDP processes through `LOCAL_RANK` and `NODE_RANK` environment variable assignments ([#7480](https://github.com/PyTorchLightning/pytorch-lightning/pull/7480))
 
 
+- Added `convert_on_fit_end` argument to `QuantizationAwareTraining` ([#8464](https://github.com/PyTorchLightning/pytorch-lightning/pull/8464))
+
+
 ### Changed
 
 

--- a/pytorch_lightning/callbacks/quantization.py
+++ b/pytorch_lightning/callbacks/quantization.py
@@ -124,7 +124,7 @@ class QuantizationAwareTraining(Callback):
             to find which layer types can be fused, check https://github.com/pytorch/pytorch/pull/43286.
 
         input_compatible: preserve quant/dequant layers. This allows to feat any input as to the original model,
-            but break compatibility to torchscript.
+            but break compatibility to torchscript and export with ``torch.save``.
 
         convert_on_fit_end: perform the quantization in `on_fit_end`.
             Note that once converted model canot be put in training again.

--- a/pytorch_lightning/callbacks/quantization.py
+++ b/pytorch_lightning/callbacks/quantization.py
@@ -98,7 +98,8 @@ class QuantizationAwareTraining(Callback):
 
             - 'fbgemm' for server inference.
             - 'qnnpack' for mobile inference.
-            -  a custom `torch.quantization.QConfig <https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig>`_.
+            -  a custom `torch.quantization.QConfig
+                <https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig>`_.
 
         observer_type: allows switching between ``MovingAverageMinMaxObserver`` as "average" (default)
             and ``HistogramObserver`` as "histogram" which is more computationally expensive.

--- a/pytorch_lightning/callbacks/quantization.py
+++ b/pytorch_lightning/callbacks/quantization.py
@@ -212,8 +212,7 @@ class QuantizationAwareTraining(Callback):
             # used with each activation tensor, fuses modules where appropriate,
             # and replaces key operators with quantized implementations.
             torch.quantization.convert(pl_module, inplace=True)
-        # check we shall preserve wrapper
-        if self._input_compatible:
-            pl_module.forward = wrap_quantize_forward_context(model=pl_module, func=self.__module_forward)
-        else:
-            pl_module.forward = self.__module_forward
+            # check we shall preserve wrapper
+            if self._input_compatible:
+                pl_module.forward = wrap_quantize_forward_context(model=pl_module, func=self.__module_forward)
+        pl_module.forward = self.__module_forward

--- a/pytorch_lightning/callbacks/quantization.py
+++ b/pytorch_lightning/callbacks/quantization.py
@@ -127,7 +127,7 @@ class QuantizationAwareTraining(Callback):
             but break compatibility to torchscript and export with ``torch.save``.
 
         convert_on_fit_end: perform the quantization in `on_fit_end`.
-            Note that once converted model canot be put in training again.
+            Note that once converted, the model cannot be put in training mode again.
 
     """
     OBSERVER_TYPES = ('histogram', 'average')
@@ -140,7 +140,7 @@ class QuantizationAwareTraining(Callback):
         modules_to_fuse: Optional[Sequence] = None,
         input_compatible: bool = True,
         convert_on_fit_end: bool = True,
-    ) -> None:  # noqa: E501
+    ) -> None:
         _valid_qconf_str = isinstance(qconfig, str) and qconfig in torch.backends.quantized.supported_engines
         if not isinstance(qconfig, QConfig) and not _valid_qconf_str:
             raise MisconfigurationException(

--- a/pytorch_lightning/callbacks/quantization.py
+++ b/pytorch_lightning/callbacks/quantization.py
@@ -205,17 +205,17 @@ class QuantizationAwareTraining(Callback):
         torch.quantization.prepare_qat(pl_module, inplace=True)
 
     def on_fit_end(self, trainer, pl_module):
-        if self._convert_on_fit_end:
-            pl_module.eval()
-            # Convert the observed model to a quantized model. This does several things:
-            # quantizes the weights, computes and stores the scale and bias value to be
-            # used with each activation tensor, fuses modules where appropriate,
-            # and replaces key operators with quantized implementations.
-            torch.quantization.convert(pl_module, inplace=True)
-            # check we shall preserve wrapper
-            if self._input_compatible:
-                pl_module.forward = wrap_quantize_forward_context(model=pl_module, func=self.__module_forward)
-            else:
-                pl_module.forward = self.__module_forward
+        if not self._convert_on_fit_end:
+            pl_module.forward = self.__module_forward
+            return
+        pl_module.eval()
+        # Convert the observed model to a quantized model. This does several things:
+        # quantizes the weights, computes and stores the scale and bias value to be
+        # used with each activation tensor, fuses modules where appropriate,
+        # and replaces key operators with quantized implementations.
+        torch.quantization.convert(pl_module, inplace=True)
+        # check we shall preserve wrapper
+        if self._input_compatible:
+            pl_module.forward = wrap_quantize_forward_context(model=pl_module, func=self.__module_forward)
         else:
             pl_module.forward = self.__module_forward

--- a/pytorch_lightning/callbacks/quantization.py
+++ b/pytorch_lightning/callbacks/quantization.py
@@ -215,4 +215,7 @@ class QuantizationAwareTraining(Callback):
             # check we shall preserve wrapper
             if self._input_compatible:
                 pl_module.forward = wrap_quantize_forward_context(model=pl_module, func=self.__module_forward)
-        pl_module.forward = self.__module_forward
+            else:
+                pl_module.forward = self.__module_forward
+        else:
+            pl_module.forward = self.__module_forward

--- a/pytorch_lightning/callbacks/quantization.py
+++ b/pytorch_lightning/callbacks/quantization.py
@@ -82,6 +82,53 @@ def _recursive_hasattr(obj: Any, attribs: str, state: bool = True) -> bool:
 
 
 class QuantizationAwareTraining(Callback):
+    """
+    Quantization allows speeding up inference and decreasing memory requirements
+    by performing computations and storing tensors at lower bitwidths
+    (such as INT8 or FLOAT16) than floating point precision.
+    We use native PyTorch API so for more information
+    see `Quantization <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>`_.
+
+    .. warning:: ``QuantizationAwareTraining`` is in beta and subject to change.
+
+
+    Args:
+
+        qconfig: quantization configuration:
+
+            - 'fbgemm' for server inference.
+            - 'qnnpack' for mobile inference.
+            -  a custom `torch.quantization.QConfig <https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig>`_.
+
+        observer_type: allows switching between ``MovingAverageMinMaxObserver`` as "average" (default)
+            and ``HistogramObserver`` as "histogram" which is more computationally expensive.
+
+        collect_quantization: count or custom function to collect quantization statistics:
+
+            - ``None`` (deafult). The quantization observer is called in each module forward
+                (useful for collecting extended statistic when useing image/data augmentation).
+            - ``int``. Use to set a fixed number of calls, starting from the beginning.
+            - ``Callable``. Custom function with single trainer argument.
+                See this example to trigger only the last epoch:
+
+                .. code-block:: python
+
+                    def custom_trigger_last(trainer):
+                        return trainer.current_epoch == (trainer.max_epochs - 1)
+
+                    QuantizationAwareTraining(collect_quantization=custom_trigger_last)
+
+        modules_to_fuse: allows you fuse a few layers together as shown in
+            `diagram <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>`_
+            to find which layer types can be fused, check https://github.com/pytorch/pytorch/pull/43286.
+
+        input_compatible: preserve quant/dequant layers. This allows to feat any input as to the original model,
+            but break compatibility to torchscript.
+
+        convert_on_fit_end: perform the quantization in `on_fit_end`.
+            Note that once converted model canot be put in training again.
+
+    """
     OBSERVER_TYPES = ('histogram', 'average')
 
     def __init__(
@@ -91,51 +138,8 @@ class QuantizationAwareTraining(Callback):
         collect_quantization: Optional[Union[int, Callable]] = None,
         modules_to_fuse: Optional[Sequence] = None,
         input_compatible: bool = True,
-    ) -> None:
-        """
-        Quantization allows speeding up inference and decreasing memory requirements
-        by performing computations and storing tensors at lower bitwidths
-        (such as INT8 or FLOAT16) than floating point precision.
-        We use native PyTorch API so for more information
-        see `Quantization <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>`_.
-
-        .. warning:: ``QuantizationAwareTraining`` is in beta and subject to change.
-
-
-        Args:
-
-            qconfig: quantization configuration:
-
-                - 'fbgemm' for server inference.
-                - 'qnnpack' for mobile inference.
-                -  a custom `torch.quantization.QConfig <https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig>`_.
-
-            observer_type: allows switching between ``MovingAverageMinMaxObserver`` as "average" (default)
-                and ``HistogramObserver`` as "histogram" which is more computationally expensive.
-
-            collect_quantization: count or custom function to collect quantization statistics:
-
-                - ``None`` (deafult). The quantization observer is called in each module forward
-                    (useful for collecting extended statistic when useing image/data augmentation).
-                - ``int``. Use to set a fixed number of calls, starting from the beginning.
-                - ``Callable``. Custom function with single trainer argument.
-                    See this example to trigger only the last epoch:
-
-                    .. code-block:: python
-
-                        def custom_trigger_last(trainer):
-                            return trainer.current_epoch == (trainer.max_epochs - 1)
-
-                        QuantizationAwareTraining(collect_quantization=custom_trigger_last)
-
-            modules_to_fuse: allows you fuse a few layers together as shown in
-                `diagram <https://pytorch.org/docs/stable/quantization.html#quantization-aware-training>`_
-                to find which layer types can be fused, check https://github.com/pytorch/pytorch/pull/43286.
-
-            input_compatible: preserve quant/dequant layers. This allows to feat any input as to the original model,
-                but break compatibility to torchscript.
-
-        """  # noqa: E501
+        convert_on_fit_end: bool = True,
+    ) -> None:  # noqa: E501
         _valid_qconf_str = isinstance(qconfig, str) and qconfig in torch.backends.quantized.supported_engines
         if not isinstance(qconfig, QConfig) and not _valid_qconf_str:
             raise MisconfigurationException(
@@ -157,6 +161,7 @@ class QuantizationAwareTraining(Callback):
 
         self.modules_to_fuse = modules_to_fuse
         self._input_compatible = input_compatible
+        self._convert_on_fit_end = convert_on_fit_end
         self._forward_calls = 0
 
     def _check_feasible_fuse(self, model):
@@ -199,12 +204,13 @@ class QuantizationAwareTraining(Callback):
         torch.quantization.prepare_qat(pl_module, inplace=True)
 
     def on_fit_end(self, trainer, pl_module):
-        pl_module.eval()
-        # Convert the observed model to a quantized model. This does several things:
-        # quantizes the weights, computes and stores the scale and bias value to be
-        # used with each activation tensor, fuses modules where appropriate,
-        # and replaces key operators with quantized implementations.
-        torch.quantization.convert(pl_module, inplace=True)
+        if self._convert_on_fit_end:
+            pl_module.eval()
+            # Convert the observed model to a quantized model. This does several things:
+            # quantizes the weights, computes and stores the scale and bias value to be
+            # used with each activation tensor, fuses modules where appropriate,
+            # and replaces key operators with quantized implementations.
+            torch.quantization.convert(pl_module, inplace=True)
         # check we shall preserve wrapper
         if self._input_compatible:
             pl_module.forward = wrap_quantize_forward_context(model=pl_module, func=self.__module_forward)

--- a/pytorch_lightning/callbacks/quantization.py
+++ b/pytorch_lightning/callbacks/quantization.py
@@ -126,7 +126,7 @@ class QuantizationAwareTraining(Callback):
         input_compatible: preserve quant/dequant layers. This allows to feat any input as to the original model,
             but break compatibility to torchscript and export with ``torch.save``.
 
-        convert_on_fit_end: perform the quantization in `on_fit_end`.
+        quantize_on_fit_end: perform the quantization in `on_fit_end`.
             Note that once converted, the model cannot be put in training mode again.
 
     """
@@ -139,7 +139,7 @@ class QuantizationAwareTraining(Callback):
         collect_quantization: Optional[Union[int, Callable]] = None,
         modules_to_fuse: Optional[Sequence] = None,
         input_compatible: bool = True,
-        convert_on_fit_end: bool = True,
+        quantize_on_fit_end: bool = True,
     ) -> None:
         _valid_qconf_str = isinstance(qconfig, str) and qconfig in torch.backends.quantized.supported_engines
         if not isinstance(qconfig, QConfig) and not _valid_qconf_str:
@@ -162,7 +162,7 @@ class QuantizationAwareTraining(Callback):
 
         self.modules_to_fuse = modules_to_fuse
         self._input_compatible = input_compatible
-        self._convert_on_fit_end = convert_on_fit_end
+        self._convert_on_fit_end = quantize_on_fit_end
         self._forward_calls = 0
 
     def _check_feasible_fuse(self, model):

--- a/pytorch_lightning/callbacks/quantization.py
+++ b/pytorch_lightning/callbacks/quantization.py
@@ -98,8 +98,8 @@ class QuantizationAwareTraining(Callback):
 
             - 'fbgemm' for server inference.
             - 'qnnpack' for mobile inference.
-            -  a custom `torch.quantization.QConfig
-                <https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig>`_.
+            - a custom `torch.quantization.QConfig
+              <https://pytorch.org/docs/stable/torch.quantization.html#torch.quantization.QConfig>`_.
 
         observer_type: allows switching between ``MovingAverageMinMaxObserver`` as "average" (default)
             and ``HistogramObserver`` as "histogram" which is more computationally expensive.

--- a/pytorch_lightning/trainer/trainer.py
+++ b/pytorch_lightning/trainer/trainer.py
@@ -1197,8 +1197,10 @@ class Trainer(
         return output
 
     def _parse_devices(
-        self, gpus: Optional[Union[List[int], str, int]], auto_select_gpus: bool, tpu_cores: Optional[Union[List[int],
-                                                                                                            str, int]]
+        self,
+        gpus: Optional[Union[List[int], str, int]],
+        auto_select_gpus: bool,
+        tpu_cores: Optional[Union[List[int], str, int]],
     ) -> Tuple[Optional[List[int]], Optional[Union[List[int], int]]]:
         if auto_select_gpus and isinstance(gpus, int):
             gpus = pick_multiple_gpus(gpus)

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -62,6 +62,15 @@ def test_quantization(tmpdir, observe: str, fuse: bool):
     # test that the test score is almost the same as with pure training
     assert torch.allclose(org_score, quant_score, atol=0.45)
 
+    qmodel2 = RegressionModel.load_from_checkpoint(str(trainer.checkpoint_callback.last_model_path))
+
+    trainer_args.update(dict(max_epochs=2))
+    trainer = Trainer(**trainer_args)
+    trainer.fit(qmodel2, datamodule=dm)
+
+    trainer = Trainer(callbacks=[qcb], **trainer_args)
+    trainer.fit(qmodel2, datamodule=dm)
+
 
 @RunIf(quantization=True)
 def test_quantize_torchscript(tmpdir):

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -49,7 +49,7 @@ def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
     org_score = torch.mean(torch.tensor([mean_relative_error(model(x), y) for x, y in dm.test_dataloader()]))
 
     fusing_layers = [(f'layer_{i}', f'layer_{i}a') for i in range(3)] if fuse else None
-    qcb = QuantizationAwareTraining(observer_type=observe, modules_to_fuse=fusing_layers, convert_on_fit_end=convert)
+    qcb = QuantizationAwareTraining(observer_type=observe, modules_to_fuse=fusing_layers, quantize_on_fit_end=convert)
     trainer = Trainer(callbacks=[qcb], **trainer_args)
     trainer.fit(qmodel, datamodule=dm)
 

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -93,8 +93,6 @@ def test_quantize_torchscript(tmpdir):
     trainer = Trainer(callbacks=[qcb], default_root_dir=tmpdir, max_epochs=1)
     trainer.fit(qmodel, datamodule=dm)
 
-    torch.save(qmodel, os.path.join(tmpdir, 'quantized-model.pt'))
-
     batch = iter(dm.test_dataloader()).next()
     qmodel(qmodel.quant(batch[0]))
 

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -28,15 +28,16 @@ from tests.helpers.simple_models import RegressionModel
 
 @pytest.mark.parametrize("observe", ['average', 'histogram'])
 @pytest.mark.parametrize("fuse", [True, False])
+@pytest.mark.parametrize("convert", [True, False])
 @RunIf(quantization=True)
-def test_quantization(tmpdir, observe: str, fuse: bool):
+def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
     """Parity test for quant model"""
     seed_everything(42)
     dm = RegressDataModule()
     trainer_args = dict(
         default_root_dir=tmpdir,
-        max_epochs=10,
-        gpus=1 if torch.cuda.is_available() else None,
+        max_epochs=7,
+        gpus=torch.cuda.device_count(),
     )
     model = RegressionModel()
     qmodel = copy.deepcopy(model)
@@ -47,26 +48,34 @@ def test_quantization(tmpdir, observe: str, fuse: bool):
     org_score = torch.mean(torch.tensor([mean_relative_error(model(x), y) for x, y in dm.test_dataloader()]))
 
     fusing_layers = [(f'layer_{i}', f'layer_{i}a') for i in range(3)] if fuse else None
-    qcb = QuantizationAwareTraining(observer_type=observe, modules_to_fuse=fusing_layers)
+    qcb = QuantizationAwareTraining(observer_type=observe, modules_to_fuse=fusing_layers, convert_on_fit_end=convert)
     trainer = Trainer(callbacks=[qcb], **trainer_args)
     trainer.fit(qmodel, datamodule=dm)
 
     quant_calls = qcb._forward_calls
     assert quant_calls == qcb._forward_calls
+    quant_score = torch.mean(torch.tensor([mean_relative_error(qmodel(x), y) for x, y in dm.test_dataloader()]))
+    # test that the test score is almost the same as with pure training
+    assert torch.allclose(org_score, quant_score, atol=0.35)
+    model_path = str(trainer.checkpoint_callback.best_model_path)
+
+    trainer_args.update(dict(max_epochs=1, checkpoint_callback=False))
+    if not convert:
+        trainer = Trainer(**trainer_args)
+        trainer.fit(qmodel, datamodule=dm)
+        qmodel.eval()
+        torch.quantization.convert(qmodel, inplace=True)
 
     quant_size = qmodel.model_size
-    quant_score = torch.mean(torch.tensor([mean_relative_error(qmodel(x), y) for x, y in dm.test_dataloader()]))
     # test that the trained model is smaller then initial
     size_ratio = quant_size / org_size
     assert size_ratio < 0.65
-    # test that the test score is almost the same as with pure training
-    assert torch.allclose(org_score, quant_score, atol=0.45)
 
-    trainer_args.update(dict(max_epochs=2))
-    trainer = Trainer(**trainer_args)
-    trainer.fit(qmodel, datamodule=dm)
+    # todo: make it work also with strict loading
+    qmodel2 = RegressionModel.load_from_checkpoint(model_path, strict=False)
+    # quant2_score = torch.mean(torch.tensor([mean_relative_error(qmodel2(x), y) for x, y in dm.test_dataloader()]))
+    # assert torch.allclose(quant_score, quant2_score, atol=0.01)
 
-    qmodel2 = RegressionModel.load_from_checkpoint(str(trainer.checkpoint_callback.best_model_path), strict=False)
     trainer = Trainer(**trainer_args)
     trainer.fit(qmodel2, datamodule=dm)
 

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -62,7 +62,7 @@ def test_quantization(tmpdir, observe: str, fuse: bool):
     # test that the test score is almost the same as with pure training
     assert torch.allclose(org_score, quant_score, atol=0.45)
 
-    qmodel2 = RegressionModel.load_from_checkpoint(str(trainer.checkpoint_callback.last_model_path))
+    qmodel2 = RegressionModel.load_from_checkpoint(str(trainer.checkpoint_callback.best_model_path))
 
     trainer_args.update(dict(max_epochs=2))
     trainer = Trainer(**trainer_args)

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -37,7 +37,7 @@ def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
     trainer_args = dict(
         default_root_dir=tmpdir,
         max_epochs=7,
-        gpus=torch.cuda.device_count(),
+        gpus=int(torch.cuda.is_available()),
     )
     model = RegressionModel()
     qmodel = copy.deepcopy(model)

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -78,7 +78,6 @@ def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
     assert torch.allclose(org_score, quant2_score, atol=0.45)
 
 
-
 @RunIf(quantization=True)
 def test_quantize_torchscript(tmpdir):
     """Test converting to torchscipt """

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -77,11 +77,6 @@ def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
     quant2_score = torch.mean(torch.tensor([mean_relative_error(qmodel2(x), y) for x, y in dm.test_dataloader()]))
     assert torch.allclose(org_score, quant2_score, atol=0.45)
 
-    trainer = Trainer(**trainer_args)
-    trainer.fit(qmodel2, datamodule=dm)
-
-    trainer = Trainer(callbacks=[QuantizationAwareTraining()], **trainer_args)
-    trainer.fit(qmodel2, datamodule=dm)
 
 
 @RunIf(quantization=True)

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -12,7 +12,6 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import copy
-import os
 from typing import Callable, Union
 
 import pytest

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -12,6 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import copy
+import os
 from typing import Callable, Union
 
 import pytest
@@ -56,12 +57,12 @@ def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
     assert quant_calls == qcb._forward_calls
     quant_score = torch.mean(torch.tensor([mean_relative_error(qmodel(x), y) for x, y in dm.test_dataloader()]))
     # test that the test score is almost the same as with pure training
-    assert torch.allclose(org_score, quant_score, atol=0.35)
-    model_path = str(trainer.checkpoint_callback.best_model_path)
+    assert torch.allclose(org_score, quant_score, atol=0.4)
+    model_path = trainer.checkpoint_callback.best_model_path
 
     trainer_args.update(dict(max_epochs=1, checkpoint_callback=False))
     if not convert:
-        trainer = Trainer(**trainer_args)
+        trainer = Trainer(callbacks=[QuantizationAwareTraining()], **trainer_args)
         trainer.fit(qmodel, datamodule=dm)
         qmodel.eval()
         torch.quantization.convert(qmodel, inplace=True)
@@ -73,13 +74,13 @@ def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
 
     # todo: make it work also with strict loading
     qmodel2 = RegressionModel.load_from_checkpoint(model_path, strict=False)
-    # quant2_score = torch.mean(torch.tensor([mean_relative_error(qmodel2(x), y) for x, y in dm.test_dataloader()]))
-    # assert torch.allclose(quant_score, quant2_score, atol=0.01)
+    quant2_score = torch.mean(torch.tensor([mean_relative_error(qmodel2(x), y) for x, y in dm.test_dataloader()]))
+    assert torch.allclose(org_score, quant2_score, atol=0.4)
 
     trainer = Trainer(**trainer_args)
     trainer.fit(qmodel2, datamodule=dm)
 
-    trainer = Trainer(callbacks=[qcb], **trainer_args)
+    trainer = Trainer(callbacks=[QuantizationAwareTraining()], **trainer_args)
     trainer.fit(qmodel2, datamodule=dm)
 
 
@@ -91,6 +92,8 @@ def test_quantize_torchscript(tmpdir):
     qcb = QuantizationAwareTraining(input_compatible=False)
     trainer = Trainer(callbacks=[qcb], default_root_dir=tmpdir, max_epochs=1)
     trainer.fit(qmodel, datamodule=dm)
+
+    torch.save(qmodel, os.path.join(tmpdir, 'quantized-model.pt'))
 
     batch = iter(dm.test_dataloader()).next()
     qmodel(qmodel.quant(batch[0]))

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -62,9 +62,11 @@ def test_quantization(tmpdir, observe: str, fuse: bool):
     # test that the test score is almost the same as with pure training
     assert torch.allclose(org_score, quant_score, atol=0.45)
 
-    qmodel2 = RegressionModel.load_from_checkpoint(str(trainer.checkpoint_callback.best_model_path))
-
     trainer_args.update(dict(max_epochs=2))
+    trainer = Trainer(**trainer_args)
+    trainer.fit(qmodel, datamodule=dm)
+
+    qmodel2 = RegressionModel.load_from_checkpoint(str(trainer.checkpoint_callback.best_model_path), strict=False)
     trainer = Trainer(**trainer_args)
     trainer.fit(qmodel2, datamodule=dm)
 

--- a/tests/callbacks/test_quantization.py
+++ b/tests/callbacks/test_quantization.py
@@ -57,7 +57,7 @@ def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
     assert quant_calls == qcb._forward_calls
     quant_score = torch.mean(torch.tensor([mean_relative_error(qmodel(x), y) for x, y in dm.test_dataloader()]))
     # test that the test score is almost the same as with pure training
-    assert torch.allclose(org_score, quant_score, atol=0.4)
+    assert torch.allclose(org_score, quant_score, atol=0.45)
     model_path = trainer.checkpoint_callback.best_model_path
 
     trainer_args.update(dict(max_epochs=1, checkpoint_callback=False))
@@ -75,7 +75,7 @@ def test_quantization(tmpdir, observe: str, fuse: bool, convert: bool):
     # todo: make it work also with strict loading
     qmodel2 = RegressionModel.load_from_checkpoint(model_path, strict=False)
     quant2_score = torch.mean(torch.tensor([mean_relative_error(qmodel2(x), y) for x, y in dm.test_dataloader()]))
-    assert torch.allclose(org_score, quant2_score, atol=0.4)
+    assert torch.allclose(org_score, quant2_score, atol=0.45)
 
     trainer = Trainer(**trainer_args)
     trainer.fit(qmodel2, datamodule=dm)

